### PR TITLE
Make colocations be compatible with DistributionStrategy in SyncReplicasOptimizer

### PR DESCRIPTION
**Pull request by wangsiyu**
This is a compatibility improvement related to `DistributionStrategy`.  The `ops.colocate_with` in 'SyncReplicasOptimizer' should be replaced with `distribution.colocate_with_vars` to support `DistributionStrategy`. Otherwise, this `colocation` will be ignored and the `self._local_step` will be placed on `PS`.

This feature can be reffered to `parameter_server_strategy_test.py`.

```
# The device scope is ignored for variables but not for normal ops.
with ops.device('/job:worker/task:0'):
  x = variable_scope.get_variable(
        'x', initializer=10.0,
         aggregation=variable_scope.VariableAggregation.SUM)
```

```
# The colocate_vars_with can override the distribution's device.
with d.colocate_vars_with(x):
  y = variable_scope.get_variable(
        'y', initializer=20.0,
        aggregation=variable_scope.VariableAggregation.SUM)
```

### CODE REVIEW

**Comment by wangsiyu** [comment]
Hi @yuefengz .
  This is a small code change related to `SyncReplicasOptimizer`. Please check it. 
Thanks.

**Comment by wangsiyu** [comment]
Hi, @rmlarsen 
  These UT failures seem not related to me. Could you start it again?
Thanks.

**Comment by wangsiyu** [comment]
Hi @rmlarsen @yuefengz ,
  All CI test has passed. Could you please help review this PR? It has been pending for long time.
Thanks.

**Comment by anj-s** [comment]
@wangsiyu Sorry for the delayed response. What is the use case for using SyncReplicaOptimizer with DistributionStrategy? You should be able to use a TensorFlow optimizer with DistributionStrategy without requiring to wrap it in a SyncReplicasOptimizer.

**Comment by wangsiyu** [comment]
@anj-s Thank you for response. My use case is to use `ParameterServerStrategy` with `sync` mode between workers. So the `SyncReplicasOptimizer` should be inserted into the model to wrap the original optimizer. Is there any other way to use `sync` mode between workers in `DistributionStrategy`?

**Comment by yuefengz** [comment]
@wangsiyu I guess this should work but I am not sure @anj-s is planning anything for colocation. Also prefer to use CollectiveAllReduceStrategy for synchronous training.

**Comment by wangsiyu** [comment]
@yuefengz `CollectiveAllReduceStrategy ` is in sync mode but it is not based on Parameter Server architecture. Sometimes the users pick `ParameterServerStrategy` for distributing the training variables across servers.

**Comment by anj-s** [comment]
@wangsiyu We currently don't support sync mode with ParameterServerStrategy. It is definitely a use case that we want to support in the future but we are not actively working on it. We want to explore other solutions that don't depend on SyncReplicaOptimizer going forward. 

Does the CollectiveAllReduceStrategy not work for your use case? What issues do you run into?

**Comment by wangsiyu** [comment]
@anj-s This is because many businesses use sync mode based on Parameter Server architecture in Alibaba including CNN and Seq2Seq models. CNN may works well with `CollectiveAllReduceStrategy ` because we usually don't need to partition variables across different servers. However, some Seq2Seq tasks may involved with large variables but update in sparse.  Although async mode is also suitable for computation in this case but sync mode may be better for convergence. 

You mentioned that you are not actively working on sync mode in PS strategy and will explore other solutions which is not depend on `SyncReplicasOptimizer`. Does it mean PS strategy will deprecat `SyncReplicasOptimizer` in the future? Why ? Is it because we cannot avoid modifying models code to wrap common optimizer？

**Review by yuefengz — APPROVED** [review]

**Comment by yuefengz** [comment]
This change looks good to me and we will have a better story for `SyncReplicaOptimizer` soon.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
